### PR TITLE
performance tests

### PR DESCRIPTION
**Pull request by snopoke**
@dannyroberts @czue FYI

https://docs.google.com/a/dimagi.com/spreadsheets/d/17ROce9wpmFs5TuNIoBnXMpTreWUynd4QRGAFCX4LEso/edit#gid=0

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/tests/__init__.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
         OtaRestoreTestBatched,
         BatchRestoreTests,
     )
+    from .performance_tests import SyncPerformanceTest
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'SyncPerformanceTest' imported but unused

File: corehq/ex-submodules/casexml/apps/phone/tests/performance_tests.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                case_type=REFERRAL_TYPE,
+                version=V2,
+                index={'parent': (PARENT_TYPE, parent_case_id)}
+            ).as_xml()
+            self._postFakeWithSyncToken(caseblock, self.sync_log.get_id)
+
+        all_cases = id_list + new_case_ids
+        assert_user_has_cases(self, self.referral_user, all_cases, restore_id=self.referral_sync_log.get_id)
+
+    @line_profile([
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/performance_tests.py]: F811 redefinition of unused 'test_profile_get_related_cases' from line 80

**Comment by czue** [comment]
very cool.

we probably don't want to run these on every build, right?

**Comment by snopoke** [comment]
Definitely not since they aren't 'real' tests. Just putting them out there
in case anyone wanted to see where the numbers were coming from.

**Comment by czue** [comment]
@snopoke seems like we should just comment out the import and merge these so they can be run manually by anyone who wants to?

**Comment by snopoke** [comment]
Yup will do

On 22 March 2015 at 20:46, Cory Zue <notifications@github.com> wrote:

> @snopoke <https://github.com/snopoke> seems like we should just comment
> out the import and merge these so they can be run manually by anyone who
> wants to?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5945#issuecomment-84672674>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

File: corehq/ex-submodules/casexml/apps/phone/tests/performance_tests.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            caseblocks.append(CaseBlock(
+                create=True,
+                case_id=case_id,
+                user_id=USER_ID,
+                owner_id=REFERRED_TO_GROUP,
+                case_type=REFERRAL_TYPE,
+                version=V2,
+                index={'parent': (PARENT_TYPE, parent_case_id)}
+            ).as_xml())
+        self._postFakeWithSyncToken(caseblocks, self.sync_log.get_id)
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/performance_tests.py]: @czue changing these to post in bulk (all cases in one form) causes the test to fail. Oddly enough if the parent cases are all created in a single form then the test passes but if the parent cases are created one case per form them when this form is processed it complains that the parent cases don't exist. 

I don't have time to look at this now but thought you might have an idea?

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/tests/performance_tests.py]: Nothing jumps out but happy to take a look. I have to imagine we support creating a parent and children in the same form.... right? Would be crazy if that's never come up till now.

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/tests/performance_tests.py]: finally took a look at this. was silly https://github.com/dimagi/commcare-hq/commit/fa4b50a466853eb11ab9288381096ba6e9566e30

**Comment by czue** [comment]
@snopoke I think this is good to merge once tests pass

**Comment by czue** [comment]
build is sad:
```
======================================================================
FAIL: testIndexesSync (casexml.apps.phone.tests.test_sync_mode.MultiUserSyncTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py", line 730, in testIndexesSync
    purge_restore_cache=True)
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 85, in assert_user_has_case
    return assert_user_has_cases(testcase, user, [case_id], return_single=True, **kwargs)
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 91, in assert_user_has_cases
    should_have=True, line_by_line=False, version=V2, **kwargs)
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 164, in check_user_has_case
    matches = [check_block(case_block) for case_block in case_blocks]
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 160, in check_block
    % (case_id, user.username, extra_info()))
AssertionError: Block for case_id 'indexes_sync_parent' doesn't appear in ota restore for user 'ferrel':
<ns0:case xmlns:ns0="http://commcarehq.org/case/transaction/v2" case_id="indexes_sync_parent" date_modified="2015-03-31" />
[]
```
I'll take a look probably tomorrow unless @snopoke beats me to it

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 4]
```diff
@@ -70,8 +72,12 @@ def _postWithSyncToken(self, filename, token_id):
         process_cases(form)
         return form
 
-    def _postFakeWithSyncToken(self, caseblock, token_id):
-        return post_case_blocks([caseblock], form_extras={"last_sync_token": token_id})
+    def _postFakeWithSyncToken(self, caseblocks, token_id):
+        if not isinstance(caseblocks, list):
+            # can't use list(caseblocks) since that returns children of the node
+            # http://lxml.de/tutorial.html#elements-are-lists
+            caseblocks = [caseblocks]
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: @czue this should fix it. I was previously doing `list(caseblocks)`

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: sneaky

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: is this going to break if you pass in a list? seems like that would result in `[[caseblocks]] `

Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: `if not isinstance(caseblocks, list)` should prevent that shouldn't it?

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: sorry, yes. i didn't read the full diff

**Pull request merged**